### PR TITLE
fix (libzkp): check tx num when CCC ErrUnknown error occurred

### PR DESCRIPTION
**Pull request by silathdiir**
## 1. Purpose or design rationale of this PR

Use CCC function `get_tx_num` to check tx count when CCC ErrUnknown error occurred.

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [ ] feat: A new feature
- [x] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [x] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: miner/worker.go [hunk 1]
```diff
@@ -1145,6 +1145,17 @@ loop:
 			// However, after `ErrUnknown`, ccc might remain in an
 			// inconsistent state, so we cannot pack more transactions.
 			circuitCapacityReached = true
+			txNumInCcc, err := w.circuitCapacityChecker.GetTxNum()
```
Comment (Thegaram) [inline miner/worker.go]: Is `txNumInCcc` the count before this last (failing) transaction? Or does it include this transaction?

File: miner/worker.go [hunk 2]
```diff
@@ -1145,6 +1145,17 @@ loop:
...
+			if err == nil {
+				txNumInMiner := uint64(w.current.tcount)
+				if txNumInMiner == txNumInCcc {
+					log.Info("tx count in miner is same with CCC", "txNum", txNumInMiner)
```
Comment (Thegaram) [inline miner/worker.go]:
```suggestion
```

No need to log in the expected case

File: miner/worker.go [hunk 3]
```diff
@@ -1164,6 +1175,17 @@ loop:
 			// inconsistent state, so we cannot pack more transactions.
 			w.eth.TxPool().RemoveTx(tx.Hash(), true)
 			circuitCapacityReached = true
+			txNumInCcc, err := w.circuitCapacityChecker.GetTxNum()
```
Comment (Thegaram) [inline miner/worker.go]: I think we should merge these two `case` branches into one to reduce code duplication. (Ping me if you need me to do this)

File: rollup/circuitcapacitychecker/mock.go [hunk 4]
```diff
@@ -43,6 +43,10 @@ func (ccc *CircuitCapacityChecker) ApplyBlock(traces *types.BlockTrace) (*types.
 	}}, nil
 }
 
+func (ccc *CircuitCapacityChecker) GetTxNum() (uint64, error) {
+	return 0, nil
```
Comment (Thegaram) [inline rollup/circuitcapacitychecker/mock.go]: We'll see constant errors when running with mock ccc

File: rollup/circuitcapacitychecker/impl.go [hunk 5]
```diff
@@ -141,3 +141,31 @@ func (ccc *CircuitCapacityChecker) ApplyBlock(traces *types.BlockTrace) (*types.
...
+	result := &WrappedTxNum{}
+	if err = json.Unmarshal([]byte(C.GoString(rawResult)), result); err != nil {
+		log.Error("fail to json unmarshal get_tx_num result", "id", ccc.ID, "err", err)
+		return nil, ErrUnknown
+	}
+	if result.Error != "" {
+		log.Error("fail to get_tx_num in CircuitCapacityChecker", "id", ccc.ID, "err", result.Error)
+		return nil, ErrUnknown
+	}
+	if result.TxNum == nil {
```
Comment (Thegaram) [inline rollup/circuitcapacitychecker/impl.go]: `TxNum` is `uint64`, it cannot be `nil`

File: miner/worker.go (hunk 3 above)
Comment (silathdiir) [reply miner/worker.go]: Add a function [checkCurrentTxNum](https://github.com/scroll-tech/go-ethereum/pull/505/commits/6387d3f2b22e0928ff7ab62e53742ab687457747#diff-689874b30f6f905ce6c47cdefeddcf052b467a7936a981f4b6cf38939e10d924R1213) in commit [6387d3f](https://github.com/scroll-tech/go-ethereum/pull/505/commits/6387d3f2b22e0928ff7ab62e53742ab687457747).

File: rollup/circuitcapacitychecker/mock.go (hunk 4 above)
Comment (silathdiir) [reply rollup/circuitcapacitychecker/mock.go]: Fix to return [uint64(ccc.countdown)](https://github.com/scroll-tech/go-ethereum/pull/505/commits/6387d3f2b22e0928ff7ab62e53742ab687457747#diff-914b7258328693ea76e3080bf0d057bdf98257e3651a24ba9f8b2773aef436f7R47) in commit [6387d3f](https://github.com/scroll-tech/go-ethereum/pull/505/commits/6387d3f2b22e0928ff7ab62e53742ab687457747).

File: rollup/circuitcapacitychecker/impl.go (hunk 5 above)
Comment (silathdiir) [reply rollup/circuitcapacitychecker/impl.go]: Delete this check and update `TxNum` to type uint64 in commit [6387d3f](https://github.com/scroll-tech/go-ethereum/pull/505/commits/6387d3f2b22e0928ff7ab62e53742ab687457747).

File: miner/worker.go (hunk 1 above)
Comment (silathdiir) [reply miner/worker.go]: I checked [this row_usages.push code of scroll-prover](https://github.com/scroll-tech/scroll-prover/blob/main/prover/src/zkevm/capacity_checker.rs#L231) as @lispc mentioned, it seems to be the count before last failed.

@lispc deleted `w.current.tcount + 1` in commit [099504c](https://github.com/scroll-tech/go-ethereum/pull/505/commits/099504c9c22152414b9cea5a166d83f4203fb6ea), please correct me if I was wrong.

File: rollup/circuitcapacitychecker/mock.go (hunk 4 above)
Comment (Thegaram) [reply rollup/circuitcapacitychecker/mock.go]: We should not use `ccc.countdown`, it is only used in tests. But when our partners build l2geth using `make nccc_geth` and run it, they will also run the mock ccc.

Comment (0xmountaintop) [reply rollup/circuitcapacitychecker/mock.go]: I will take care of this PR @silathdiir

Comment (silathdiir) [reply rollup/circuitcapacitychecker/mock.go]: I fix to add `tcount` to mock CCC and a new function `SetTxNum` to this `mock.go` in commit [a99fc44](https://github.com/scroll-tech/go-ethereum/pull/505/commits/a99fc443402b75abb3462af7ec42c6218c97fa39).

Please correct me if I misunderstood your comment, thanks.

File: miner/worker.go [hunk 6]
```diff
@@ -1208,6 +1210,18 @@ loop:
...
+	txNumInCcc, err := w.circuitCapacityChecker.GetTxNum()
+	if err == nil {
+		txNumInMiner := uint64(w.current.tcount)
+		if txNumInMiner != txNumInCcc {
+			log.Error("tx count in miner is different with CCC", "txNumInMiner", txNumInMiner, "txNumInCcc", txNumInCcc)
+		}
+	} else {
+		log.Error("failed to get_tx_num", "err", err)
+	}
+}
```
Comment (georgehao) [inline miner/worker.go]:
```suggestion
func (w *worker) checkCurrentTxNum() {
	txNumInCcc, err := w.circuitCapacityChecker.GetTxNum()
	if err != nil {
		log.Error("checkCurrentTxNum failed to GetTxNum", "err", err)
		return
	}
	
	txNumInMiner := uint64(w.current.tcount)
	if txNumInMiner != txNumInCcc {
		log.Error("tx count in miner is different with CCC", "txNumInMiner", txNumInMiner, "txNumInCcc", txNumInCcc)
	}
}
```

File: rollup/circuitcapacitychecker/impl.go [hunk 7]
```diff
@@ -141,3 +141,27 @@ func (ccc *CircuitCapacityChecker) ApplyBlock(traces *types.BlockTrace) (*types.
 	}
 	return (*types.RowConsumption)(&result.AccRowUsage.RowUsageDetails), nil
 }
+
+func (ccc *CircuitCapacityChecker) GetTxNum() (uint64, error) {
```
Comment (georgehao) [inline rollup/circuitcapacitychecker/impl.go]:
```suggestion
// GetTxNum get the tx number of ccc
func (ccc *CircuitCapacityChecker) GetTxNum() (uint64, error) {
```

File: miner/worker.go [hunk 8]
```diff
@@ -1208,6 +1210,18 @@ loop:
 	return false, circuitCapacityReached
 }
 
+func (w *worker) checkCurrentTxNum() {
+	txNumInCcc, err := w.circuitCapacityChecker.GetTxNum()
+	if err == nil {
+		txNumInMiner := uint64(w.current.tcount)
+		if txNumInMiner != txNumInCcc {
```
Comment (Thegaram) [inline miner/worker.go]: I wonder if this scenario is possible:
1. Tracing completes with no error,
2. CCC completes with no error,
3. `ApplyTransaction` fails with an error.

In this case the two transaction counts would get out of sync. But I guess if 3. fails then 1. fails too, is that correct? @HAOYUatHZ

Comment (0xmountaintop) [reply miner/worker.go]: yes. you can check `func applyTransaction` in core/state_processor.go

errors can only happen in `fees.CalculateL1DataFee` and `core.ApplyMessage`, which will also be caught in `core/traces.go`

**Comment by 0xmountaintop** [comment]
I’ve refactored this PR @georgehao @silathdiir 

1. so that we don’t need to worry about whether `checkCurrentTxNumWithCCC` will be correct in mock_ccc
2. I use
```
func (w *worker) checkCurrentTxNumWithCCC(expected int) {
	match, got, err := w.circuitCapacityChecker.CheckTxNum(expected) 
```
instead of 
```
func (w *worker) checkCurrentTxNumWithCCC() {
	match, got, err := w.circuitCapacityChecker.CheckTxNum(w.current.tcount)
```

because not sure in the future whether we will need to check (ccc_tx  ==  (w.current.tcount+1)),  for example, in ccc successful cases.

**Review by georgehao — APPROVED** [review]

**Review by Thegaram — APPROVED** [review]

**Pull request merged**